### PR TITLE
[INLONG-7058][Sort] Support Apache Kudu connector

### LINKED ISSUE #7058: [Feature][Sort] Support Apache Kudu connector
### Description

_No response_

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by featzhang**
## Prepare a Pull Request
*[INLONG-7058][Sort] Support Apache kudu connector*

- Fixes #7058 

### Motivation

```sql
CREATE TABLE kudu_tbl(
uid int,
fee double,
event_id bigint,
remark varchar
)
WITH(
 'connector'='kudu-inlong',
 'masters'='localhost:8081;localhost:8080',
 'table'='actual_tbl',
 'flush-mode'='AUTO_FLUSH_SYNC',
 'lookup.max-cache-size' = '-1',
 'lookup.max-cache-time' = '60s',
 'sink.start-new-chain' = 'true',
 'max-retries' = '3',
 'sink.max-buffer-size' = '100',
 'sink.write-thread-count' = '5', 
 'sink.max-buffer-time' = '30s',
 'sink.write-with-async-mode' = 'false',
 'sink.write-with-upsert-mode' = 'true',
 'sink.cache-queue-max-length' = '-1',
 'sink.ignore-all-changelog' = 'false'
);
```

The `Apache Kudu` `sort-connector` is fully integrated with Flink 1.13.x's Table and SQL API. By configuring `KuduTableInfo` (`connector.masters`, `connector.table` and other attributes), the kudu table can be queried and inserted by Flink SQL.

#### DDL options

| Option | Default value| Required | Remarks |
| ---- | ---- | ---- | ---- |
| `connector.masters` | null  |   true | The master addresses of kudu server, like 'localhost:8080,localhost:8081' |
| `connector.table` | null  |   true | The table name of kudu |
| `flush-mode` | AUTO_FLUSH_SYNC  |   true | The flush mode of Kudu client session, AUTO_FLUSH_SYNC/AUTO_FLUSH_BACKGROUND/MANUAL_FLUSH. |
| `lookup.max-cache-size` | -1  |   true | The maximum number of results cached in the lookup source. |
| `lookup.max-cache-time` | 60s  |   true | The maximum live time for cached results in the lookup source. |
| `sink.start-new-chain` | true  |   true | The sink operator will start a new chain if true. |
| `max-retries` | 3  |   true | The maximum number of retries when an exception is caught. |
| `sink.max-buffer-size` | 100  |   true | The maximum number of records buffered in the sink. |
| `sink.write-thread-count` | 5  |   true | The maximum number of thread in the sink. |
| `sink.max-buffer-time` | 30s  |   true | The maximum wait time for buffered records in the sink. |
| `sink.write-with-async-mode` | false  |   true | Use async write mode for kudu producer if true. |
| `sink.write-with-upsert-mode` | true  |   true | Force write kudu with upsert mode if true.  This option is used for upsert using the primary key of the kudu table |
| `sink.cache-queue-max-length` | -1  |   true | The maximum queue lengths. |
| `sink.ignore-all-changelog` | false   |  true | Whether ignore delete/update_before/update_after message. If true, all message will insert|

#### Data type

| Flink/SQL | Kudu data type |
| --- | --- |
| STRING | STRING |
| BOOLEAN | BOOL |
| TINYINT | INT8 |
| SMALLINT | INT16 |
| INT | INT32 |
| BIGINT | INT64 |
| FLOAT | FLOAT |
| DOUBLE | DOUBLE |
| BYTES | BINARY |
| TIMESTAMP(3) | UNIXTIME_MICROS |

image

<br class="Apple-interchange-newline">

### Verifying this change

*(Please pick either of the following options)*

- [x] This change is a trivial rework/code cleanup without any test coverage.

- [x] This change is already covered by existing tests, such as:
  *(please describe tests)*

- [ ] This change added tests and can be verified as follows:

  *(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a follow-up issue for adding the documentation

### CODE REVIEW

File: inlong-sort/pom.xml [hunk 1]
```diff
@@ -47,6 +47,7 @@
         <hbase.version>2.2.3</hbase.version>
         <iceberg.hive.version>2.3.7</iceberg.hive.version>
         <hudi.hive.version>2.3.7</hudi.hive.version>
+        <kudu.version>1.16.0</kudu.version>
```
Comment (dockerzhang) [inline inlong-sort/pom.xml:50]: Please move the defined version to root pom.xml

File: inlong-sort/sort-connectors/kudu/pom.xml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        <dependency>
+            <groupId>org.apache.kudu</groupId>
+            <artifactId>kudu-client</artifactId>
+            <version>${kudu.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.inlong</groupId>
+            <artifactId>sort-format-common</artifactId>
+            <version>1.5.0-SNAPSHOT</version>
```
Comment (yunqingmoswu) [inline inlong-sort/sort-connectors/kudu/pom.xml:50]: Please replace it use the variable 'project.vaersion'

**Comment by yunqingmoswu** [comment]
Please add some describe for your design and implements, thanks.

**Review by yunqingmoswu — COMMENTED** [review]
Please add the access protocol for it(add KuduLoadNode in the sort-common).

**Comment by featzhang** [comment]
> Please add the access protocol for it(add KuduLoadNode in the sort-common).

I will submit another PR.

**Comment by featzhang** [comment]
> Please add some describe for your design and implements, thanks.

I have supplemented some document information in the pr description, please check

File: inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/common/KuduUtils.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+import static org.apache.flink.util.Preconditions.checkState;
+import static org.apache.kudu.Type.*;
+
+/**
+ * The utility class for Kudu.
+ */
+public class KuduUtils {
+
+    private static final Map<Type, TypeInfo> KUDU_TYPE_2_TYPE_INFO_MAP = new HashMap<>(13);
+    private static final Map<Type, TypeInfo> KUDU_TYPE_2_DATA_TYPE_MAP = new HashMap<>(13);
```
Comment (gong) [inline inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/common/KuduUtils.java:68]: maybe 13 should modify to 13/0.75

File: inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/sink/AbstractKuduSinkFunction.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        checkError();
+        // We must store the exception caught in the flushing so that the
+        // task thread can be aware of the failure.
+        try {
+            flush();
+        } catch (IOException e) {
+            this.flushThrowable = e;
+        }
+        checkError();
+    }
```
Comment (gong) [inline inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/sink/AbstractKuduSinkFunction.java:158]: lost metric snapshot call

File: inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/common/KuduOptions.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    public static final ConfigOption<Boolean> LAZY_LOAD_SCHEMA =
+            ConfigOptions.key("lazy.load.schema")
+                    .booleanType()
+                    .defaultValue(false)
+                    .withDescription("Whether get metadata from kudu after generating flink job graphic.");
+    public static final ConfigOption<Boolean> KUDU_IGNORE_ALL_CHANGELOG =
+            ConfigOptions.key("ignore.all.changelog")
+                    .booleanType()
+                    .defaultValue(false)
+                    .withDescription("Whether ignore delete/update_before/update_after message.");
```
Comment (gong) [inline inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/common/KuduOptions.java:122]: maybe `ignore.all.changelog` change to `sink.ignore.changelog`. keep consistent with other connector

File: inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/table/KuduDynamicTableSink.java [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+    @Override
+    public ChangelogMode getChangelogMode(ChangelogMode requestedMode) {
+        if (org.apache.flink.configuration.Configuration.fromMap(catalogTable.getOptions())
+                .get(KUDU_IGNORE_ALL_CHANGELOG)) {
+            LOG.warn("Kudu sink receive all changelog record. "
+                    + "Regard any other record as insert-only record.");
+            return ChangelogMode.all();
+        }
+        return ChangelogMode.all();
```
Comment (gong) [inline inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/table/KuduDynamicTableSink.java:176]: maybe just `return ChangelogMode.all();`

File: inlong-sort/sort-connectors/kudu/pom.xml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                        <phase>package</phase>
+                        <configuration>
+                            <artifactSet>
+                                <includes>
+                                    <include>org.apache.hudi:*</include>
+                                    <include>org.apache.hive:hive-exec</include>
+                                    <include>org.apache.hadoop:*</include>
+                                    <include>com.fasterxml.woodstox:*</include>
+                                    <include>org.codehaus.woodstox:*</include>
+                                    <include>com.google.guava:*</include>
```
Comment (gong) [inline inlong-sort/sort-connectors/kudu/pom.xml:117]: hudi? I can't find kudu

File: inlong-sort/sort-connectors/kudu/pom.xml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    <name>Apache InLong - Sort-connector-kudu</name>
+
+    <properties />
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.inlong</groupId>
+            <artifactId>sort-connector-base</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
Comment (gong) [inline inlong-sort/sort-connectors/kudu/pom.xml:46]: I suggest add a shade for `sort-connector-base`

**Comment by gong** [comment]
add fileSet for kudu connector in inlong-distribution/src/main/assemblies/sort-connectors.xml

File: inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/source/KuduLookupFunction.java [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+                ArrayList<Row> rows = new ArrayList<>();
+                for (final KuduScanToken token : tokenList) {
+                    final List<LocatedTablet.Replica> replicas = token.getTablet().getReplicas();
+                    final String[] array = replicas
+                            .stream()
+                            .map(replica -> replica.getRpcHost() + ":" + replica.getRpcPort())
+                            .collect(Collectors.toList()).toArray(new String[replicas.size()]);
+                    final byte[] scanToken = token.serialize();
+                    final KuduScanner scanner = KuduScanToken.deserializeIntoScanner(scanToken, client);
+                    RowResultIterator rowIterator = scanner.nextRows();
```
Comment (zhangyifan27) [inline inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/source/KuduLookupFunction.java:248]: We can use the `KuduScannerIterator` API to process the RowResults through `scanner.iterator()`, which contains keep alive calls to ensure the scanner does not time out.

File: inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/common/KuduOptions.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigOptions;
+
+/**
+ * The configuration options for kudu sink.
+ */
+public class KuduOptions {
+
+    public static final ConfigOption<String> CONNECTOR_TABLE =
```
Comment (yunqingmoswu) [inline inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/common/KuduOptions.java:28]: It is necessary for adding a prefix with 'CONNECTOR'?

File: inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/common/KuduOptions.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            ConfigOptions.key("table")
+                    .stringType()
+                    .noDefaultValue().withDescription("The name of kudu table.");
+
+    public static final ConfigOption<String> CONNECTOR_MASTERS =
```
Comment (yunqingmoswu) [inline inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/common/KuduOptions.java:33]: It is necessary for adding a prefix with 'CONNECTOR'?

File: inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/sink/AbstractKuduSinkFunction.java [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+                .withInitDirtyBytes(metricState != null ? metricState.getMetricValue(DIRTY_BYTES_OUT) : 0L)
+                .withRegisterMetric(MetricOption.RegisteredMetric.ALL)
+                .build();
+        if (metricOption != null) {
+            sinkMetricData = new SinkMetricData(metricOption, getRuntimeContext().getMetricGroup());
+        }
+    }
+
+    @Override
+    public void invoke(
```
Comment (yunqingmoswu) [inline inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/sink/AbstractKuduSinkFunction.java:134]: It is recommended to remove the newline.

**Review by zhangyifan27 — APPROVED** [review]
LGTM

**Review by yunqingmoswu — APPROVED** [review]

**Review by dockerzhang — APPROVED** [review]

File: inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/sink/AbstractKuduSinkFunction.java [hunk 13]
```diff
@@ -145,7 +149,11 @@ public void close() throws Exception {
     }
 
     @Override
-    public void snapshotState(FunctionSnapshotContext functionSnapshotContext) {
+    public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throws Exception {
+        if (sourceMetricData != null && metricStateListState != null) {
+            MetricStateUtils.snapshotMetricStateForSourceMetricData(metricStateListState, sourceMetricData,
```
Comment (gong) [inline inlong-sort/sort-connectors/kudu/src/main/java/org/apache/inlong/sort/kudu/sink/AbstractKuduSinkFunction.java:152]: here should be sinkMetricData for kudu sink

**Review by gong — APPROVED** [review]

**Pull request merged**